### PR TITLE
populate security properties

### PR DESCRIPTION
**Pull request by elsapeng**
populates security properties
writes tests on them

fixes issue #48

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -74,10 +76,34 @@ function getData(swagger, path, operation, response, config) {
...
     formParameters: [],
-    security: security,
+    basic: null,
+    apiKey: null,
+    oauth: null,
     path: ''
   };
 
+  // deal with the security properties
+  if (security !== undefined && security.length !== 0) {
```
Comment (mohsen1) [inline index.js]: `(security && security.length !== 0)` is safer

File: index.js [hunk 2]
```diff
@@ -74,10 +76,34 @@ function getData(swagger, path, operation, response, config) {
...
+    for (securityKey in Object.keys(security)) {
+      if (Object.keys(security).hasOwnProperty(securityKey)) {
```
Comment (mohsen1) [inline index.js]: `Object.keys` returns an array. Why are you using `hasOwnProperty` on it?

File: index.js [hunk 3]
```diff
@@ -74,10 +76,34 @@ function getData(swagger, path, operation, response, config) {
...
-    security: security,
+    basic: null,
+    apiKey: null,
+    oauth: null,
     path: ''
   };
 
+  // deal with the security properties
+  if (security !== undefined && security.length !== 0) {
+    for (securityKey in Object.keys(security)) {
```
Comment (mohsen1) [inline index.js]: To loop in `Object.keys` result use `forEach`.

File: index.js [hunk 4]
```diff
@@ -74,10 +76,34 @@ function getData(swagger, path, operation, response, config) {
...
+          case 'basic':
+            data.basic = securityType;
+            break;
+          case 'apiKey':
+            data.apiKey = securityType;
+            break;
+          case 'oauth2':
+            data.oauth = securityType;
+            break;
+          default:
```
Comment (mohsen1) [inline index.js]: throw if type is weird

File: index.js [hunk 5]
```diff
@@ -131,6 +157,25 @@ function getData(swagger, path, operation, response, config) {
     }
   }
 
+  // deal with apiKey properties if it is in query
+  if (data.apiKey != null && data.apiKey.in === 'query') {
```
Comment (mohsen1) [inline index.js]: `(data.apiKey && data.apiKey.in === 'query')`

File: index.js [hunk 6]
```diff
@@ -131,6 +157,25 @@ function getData(swagger, path, operation, response, config) {
...
+    var item = {
+      name: data.apiKey.name,
+      in: 'query',
+      type: 'string'
+    };
+
+    data.queryParameters.push(item);
+  } else if (data.apiKey != null && data.apiKey.in === 'header') {
```
Comment (mohsen1) [inline index.js]: Same...

File: package.json [hunk 7]
```diff
@@ -23,7 +23,7 @@
       "url": "http://azimi.me"
     },
     {
-      "name": "Linjie Penjg",
+      "name": "Linjie Peng",
```
Comment (mohsen1) [inline package.json]: Sorry I misspelled this :(

File: templates/request/delete/delete.handlebars [hunk 8]
```diff
@@ -19,8 +19,10 @@
           'Content-Type': '{{contentType}}'{{#if headerParameters}},
           'Custom-Header': {
           {{#each headerParameters}}
-          {{this.name}}: 'DATA GOES HERE'{{#unless @last}},{{/unless}}
-          {{/each}} }{{/if}}
+            {{this.name}}: 'DATA GOES HERE'{{#unless @last}},{{/unless}}
+          {{/each}} }{{/if}}{{#if basic}},
+          Authorization: 'Basic TOKEN GOES HERE'{{/if}}{{#if oauth}},
```
Comment (mohsen1) [inline templates/request/delete/delete.handlebars]: I think we should use environment variables for secrets. People don't commit their secrets usually. As part of code generation we can have a `.env` file to hold the placeholders for the environment variables.

So here is should be something like: 

```js
Authorization: 'Basic ' + process.env.BASIC_AUTH_TOKEN
```

Comment (mohsen1) [inline templates/request/delete/delete.handlebars]: This of course applies to all types of secret

File: index.js (hunk 4 above)
Comment (elsapeng) [inline index.js]: will change to array.foreach

File: templates/request/delete/delete.handlebars (hunk 8 above)
Comment (elsapeng) [inline templates/request/delete/delete.handlebars]: got it, will change them

File: index.js [hunk 9]
```diff
@@ -74,10 +77,37 @@ function getData(swagger, path, operation, response, config) {
...
     path: ''
   };
 
+  // deal with the security properties
+  if (security && security.length !== 0) {
+    Object.keys(security).forEach(function(element) {
+      securityType = swagger.securityDefinitions[element];
+      switch (securityType.type) {
+        case 'basic':
+          data.oauth = {name: element.toUpperCase(), type: 'Basic'};
```
Comment (mohsen1) [inline index.js]: `data.headerSecurity`

File: index.js [hunk 10]
```diff
@@ -157,6 +187,10 @@ function getData(swagger, path, operation, response, config) {
         queryToAdd = element.name + '=DATA&';
         data.path += queryToAdd;
       });
+      if (data.queryApiKey) {
+        data.path += data.queryApiKey.type +
```
Comment (mohsen1) [inline index.js]: Make sure you are using a query string library for constructing this.

File: index.js [hunk 11]
```diff
@@ -367,6 +401,16 @@ function testGen(swagger, config) {
   source = read(join(__dirname, '/templates/outerDescribe.handlebars'), 'utf8');
   outerDescribeFn = handlebars.compile(source, {noEscape: true});
 
+  if (swagger.securityDefinitions) {
+    write('.env', '');
+    var keys = Object.keys(swagger.securityDefinitions);
+
+    keys.forEach(function(element) {
```
Comment (mohsen1) [inline index.js]: do this in templates

File: index.js [hunk 12]
```diff
@@ -74,10 +77,37 @@ function getData(swagger, path, operation, response, config) {
...
+            {name: element.toUpperCase(), type: securityType.name};
+          } else if (securityType.in === 'header') {
+            data.headerApiKey =
+            {name: element.toUpperCase(), type: securityType.name};
+          }
+          break;
+        case 'oauth2':
+          data.oauth = {name: element.toUpperCase(), type: 'Bearer'};
+          break;
+        default:
```
Comment (mohsen1) [inline index.js]: throw if type is weird

File: test/minimal/compare/output11-.env [hunk 13]
```diff
@@ -0,0 +1,4 @@
+/*eslint-disable*/
```
Comment (mohsen1) [inline test/minimal/compare/output11-.env]: I don't think you need to worry about eslint here.

File: test/minimal/compare/output11-base-path-test.js [hunk 14]
```diff
@@ -20,7 +20,7 @@ describe('/', function() {
     });
 
     it('should respond with 200 OK', function(done) {
-      api.get('/')
+      api.get('/?accessToken=process.env.KEY')
```
Comment (mohsen1) [inline test/minimal/compare/output11-base-path-test.js]: `=process.env.KEY` should not be part of the string. It's a variable.

File: test/minimal/compare/output11-.env (hunk 13 above)
Comment (elsapeng) [inline test/minimal/compare/output11-.env]: I actually got an eslint error here: "Use the global form of "use strict".", but this is actually not a js file

File: test/minimal/compare/output11-base-path-test.js (hunk 14 above)
Comment (elsapeng) [inline test/minimal/compare/output11-base-path-test.js]: Oh, I forget about that, will change.

File: test/minimal/compare/output11-.env (hunk 13 above)
Comment (mohsen1) [inline test/minimal/compare/output11-.env]: weird, try fixing the tests to not read `.env` files.

File: test/minimal/compare/output11-base-path-test.js (hunk 14 above)
Comment (mohsen1) [inline test/minimal/compare/output11-base-path-test.js]: `=` is extra in my comment. I meant  `process.env.KEY`

**Comment by elsapeng** [comment]
which one?

**Comment by elsapeng** [comment]
I didn't remove any rule, but let the eslint do not test the .env files

**Pull request merged**